File: typescript/repl/repl.ts
import * as readline from "readline";
import { Lexer } from "../lexer/lexer";
import { TokenType } from "../token/token";
import { Parser } from "../parser/parser";

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const PROMPT = ">> ";

export function start() {
  rl.on("line", (input) => {
    const lexer = new Lexer(input);
    const parser = new Parser(lexer);
    const program = parser.parseProgram();

    if (parser.errors.length > 0) {
      printParseErrors(parser.errors);
    }

    console.log(program.string);
    rl.prompt();
  }).on("close", () => {
    console.log("REPL terminated");
    process.exit(0);
  });

  rl.setPrompt(PROMPT);
  rl.prompt();
}

const MONKEY_FACE = `            __,__
   .--.  .-"     "-.  .--.
  / .. \/  .-. .-.  \/ .. \
 | |  '|  /   Y   \  |'  | |
 | \   \  \ 0 | 0 /  /   / |
  \ '- ,\.-"""""""-./, -' /
   ''-' /_   ^ ^   _\ '-''
       |  \._   _./  |
       \   \ '~' /   /
        '._ '-=-' _.'
           '-----'
`;

function printParseErrors(errors: string[]) {
  console.log(MONKEY_FACE);
  console.log("Whoops! We ran into some monkey business here!");
  console.log("Parser errors:");
  for (const error of errors) {
    console.log(error);
  }
}

File: python/repl/repl.py
import sys

# Assuming you have a lexer.py module with a Lexer class that has a similar interface to your Go lexer
from lexer import Lexer
from monkey_token import TokenType

PROMPT = ">> "


def start(in_stream, out_stream):
    while True:
        out_stream.write(PROMPT)
        out_stream.flush()
        try:
            line = in_stream.readline()
            if not line:  # EOF
                break
            lexer = Lexer(line)
            token = lexer.next_token()
            while token and token.type != TokenType.EOF:
                out_stream.write(f"{token}\n")
                token = lexer.next_token()
        except KeyboardInterrupt:
            out_stream.write("\nExiting...\n")
            break


if __name__ == "__main__":
    start(sys.stdin, sys.stdout)
